### PR TITLE
add docstrings to cwstring and transcode

### PR DESCRIPTION
**Pull request by simonbyrne**

### CODE REVIEW

File: base/c.jl [hunk 1]
```diff
@@ -110,8 +106,17 @@ end
...
 if is_windows()
+"""
+    Base.cwstring(s)
+
+Converts a string `s` to a NUL-terminated `Vector{Cwchar_t}`, suitable for passing to C
+functions expecting a `Ptr{Cwchar_t}`. The main advantage of using this over the implicit
+conversion provided by `Cwstring` is if the function is called multiple times with the
+same argument.
+
+This is only available on Windows.
```
Comment (Keno) [inline base/c.jl]: Should this mention that this is supposed to be used via `Cwstring`?

Comment (simonbyrne) [inline base/c.jl]: Is it? From what I can tell, the following are equivalent:
```julia
ccall(....,Cwstring, str)
```
or
```julia
cwstr = cwstring(str)
ccall(....,Ptr{Cwchar_t}, str)
```
the advantage of the 2nd one is that if you have to call the function multiple times, you only have to do the conversion once.

Comment (Keno) [inline base/c.jl]: I thought Cwstring was exported, but this was not.

File: base/c.jl [hunk 2]
```diff
@@ -75,11 +75,7 @@ cconvert(::Type{Cstring}, s::String) =
 cconvert(::Type{Cstring}, s::AbstractString) =
     cconvert(Cstring, String(s)::String)
 
-function cconvert(::Type{Cwstring}, s::AbstractString)
-    v = transcode(Cwchar_t, String(s).data)
-    !isempty(v) && v[end] == 0 || push!(v, 0)
-    return v
-end
+cconvert(::Type{Cwstring}, s::AbstractString) = cwstring(s)
```
Comment (tkelman) [inline base/c.jl]: so `cwstring` gets used everywhere but only defined on windows?

File: base/c.jl (hunk 1 above)
Comment (ararslan) [inline base/c.jl]: Looks like @Keno is right: `Cwstring` appears in base/exports.jl but not `cwstring`.

Comment (simonbyrne) [inline base/c.jl]: It isn't exported, but is that any reason not to document it?

File: base/c.jl (hunk 2 above)
Comment (simonbyrne) [inline base/c.jl]: I reverted this bit. Basically we have the same logic defined twice (once for the implicit `cconvert`, `unsafe_convert` via `Cwstring`, and once for the explicit `cwstring` function. But their precise order (i.e. when they do the checks) differs, so I decided to leave it.

**Comment by simonbyrne** [comment]
I realise these functions aren't exported, but probably good to have documented (we may want to export `transcode` at some point).

The only thing that isn't documented is the recommended way to get a `String` back from a `ccall` that returns a UTF-16 string. The examples in Base all use `String(Base.transcode(UInt8, x))`, where `x` is a `Vector{UInt16}`.

**Comment by tkelman** [comment]
Can anyone explain the justification for only defining `cwstring` on Windows?

**Comment by StefanKarpinski** [comment]
> Can anyone explain the justification for only defining cwstring on Windows?

The definition of `cwstring` doesn't work on systems where `Cwchar_t != UInt16` since we don't currently define `transcode` methods for UTF-32. I agree that this should be simplified, but that doesn't block this PR.

**Pull request merged**